src/stream/src/executor/managed_state/top_n/mod.rs

```diff
@@ -41,39 +41,63 @@ fn deserialize_pk<const TOP_N_TYPE: usize>(
     Ok(pk)
 }
 
-// The function moves `pk_row_bytes_iter` as it is possible that we would take one extra cell from
-// it than the returned result. Allowing it to be reused after this function via borrowing may cause
-// confusion and bugs.
-async fn deserialize_bytes_to_pk_and_row<
-    const TOP_N_TYPE: usize,
-    I: StateStoreIter<Item = (Bytes, Bytes)>,
->(
-    mut pk_row_bytes_iter: I,
-    ordered_row_deserializer: &mut OrderedRowDeserializer,
-    cell_based_row_deserializer: &mut CellBasedRowDeserializer,
-    num_rows: Option<usize>,
-) -> Result<Vec<(OrderedRow, Row)>> {
-    let mut result = vec![];
-    while let Some((key, value)) = pk_row_bytes_iter.next().await? {
-        let pk_buf_and_row = cell_based_row_deserializer.deserialize(&key, &value)?;
-        match pk_buf_and_row {
-            Some((mut pk_buf, row)) => {
-                let pk = deserialize_pk::<TOP_N_TYPE>(&mut pk_buf, ordered_row_deserializer)?;
-                result.push((pk, row));
-                if let Some(num_rows) = num_rows && result.len() == num_rows {
-                    return Ok(result);
+pub struct PkAndRowIterator<'a, I: StateStoreIter<Item = (Bytes, Bytes)>, const TOP_N_TYPE: usize> {
+    iter: I,
+    ordered_row_deserializer: &'a mut OrderedRowDeserializer,
+    cell_based_row_deserializer: &'a mut CellBasedRowDeserializer,
+}
```

Comment:
maybe this can be part of `CellBasedTable` and replace the `CellBasedTableRowIter`? and after that, we can implement the iter of `StateTable` composited with `MemTable`(flush_buffer) and `CellBasedTable`. and all stream operators can use the `StateTable` to store their states.

Reply:
It seems that only `TopNExecutor` requires that the `next` of the iterator returns both key and value, and the key is an `OrderedRow`. This is not required in other stateful executors, i.e. `HashAgg` and `HashJoin`. It is also not required in MV. 

This is because only `TopNExecutor` needs an ordered key to fill in its cache, and this filling process does not only rely on the key of the single`(insert, row)` whose state is being updated.

Use a flag to denote whether it should deserialize and return the key? 🤣

Reply:
Considering we will do `store a column either in key or value`, the caller should just make the decision whether it is necessary to return `key`. This iterator should figure out by itself whether it needs to deserialize `key`.

Reply:
Maybe we merge first and take some time to decide how to merge this with `CellBasedTable` and replace the `CellBasedTableRowIter`?  Some other fixes are pending in `TopN`.